pageserver/src/object_repository.rs

```diff
@@ -622,141 +623,65 @@ impl ObjectTimeline {
-                            // do not remove last version
-                            if last_lsn > horizon {
-                                // locate most recent record before horizon
-                                let key = StorageKey::des(iter.key().unwrap());
-                                if key.tag == maxkey.tag {
-                                    let val = StorageValue::des(iter.value().unwrap());
-                                    if let StorageValueContent::Image(_) = val.content {
-                                        let (base_img, records) =
-                                            self.collect_records_for_apply(key.tag, key.lsn);
-                                        trace!("Reconstruct horizon page {} blk {} at {} from {} records",
-                                              key.tag.rel, key.tag.blknum, key.lsn, records.len());
-                                        let new_img = self
-                                            .walredo_mgr
-                                            .request_redo(key.tag, key.lsn, base_img, records)?;
-                                        self.put_page_image(key.tag, key.lsn, new_img.clone());
-
-                                        truncated += 1;
-                                    } else {
-                                        trace!(
-                                            "Keeping horizon page {} blk {} at {}",
-                                            key.tag.rel,
-                                            key.tag.blknum,
-                                            key.lsn
-                                        );
-                                    }
-                                }
-                            } else {
-                                trace!(
-                                    "Last page {} blk {} at {}, horizon {}",
-                                    key.tag.rel,
-                                    key.tag.blknum,
-                                    key.lsn,
-                                    horizon
-                                );
+                // Iterate through all relations
+                for rels in &self.obj_store.list_rels(self.timelineid, 0, 0, last_lsn)? {
+                    let mut last_version = true;
+                    let mut key = relation_size_key(self.timelineid, *rels);
+                    let mut max_size = 0u32;
+                    let mut relation_dropped = false;
+
+                    // Process relation metadata versions
+                    for vers in self.obj_store.object_versions(&key, horizon)? {
+                        let lsn = vers.0;
+                        let rel_meta = RelationSizeEntry::des(&vers.1)?;
+                        if let RelationSizeEntry::Size(size) = rel_meta {
+                            max_size = max(max_size, size);
+                        } else if last_version {
+                            relation_dropped = true;
```

Comment:
It took me a while to understand how this works. It's assuming that if the entry is not of type `RelationSizeEntry::Size`, then it must be `RelationSizeEntry::Unlink`. That's very subtle. Please use the match-construct to make that more clear. Some comments would be nice too.

We don't actually parse relation drops from commit/abort records currently, so I believe this is dead code until we add that...

Reply:
> It took me a while to understand how this works. It's assuming that if the entry is not of type `RelationSizeEntry::Size`, then it must be `RelationSizeEntry::Unlink`. That's very subtle. Please use the match-construct to make that more clear. Some comments would be nice too.

Sorry, done.

> 
> We don't actually parse relation drops from commit/abort records currently, so I believe this is dead code until we add that...

I still hoping that we will sometime switch to wal-redo version where dropping relation is supported. Actually it is one more argument against replaying non-rel WAL: it is not clear how to handle drop relations in this case.